Constify `copy` related functions

usbalbin opened the pull request:
Constify 

* `*const T::copy_to[_nonoverlapping]`
* `*mut T::copy_to[_nonoverlapping]`
* `*mut T::copy_from[_nonoverlapping]`
* `mem::transmute_copy`
* `mem::swap`
* `ptr::swap[_nonoverlapping]`
* `mem::replace`
* `ptr::replace`

rust-highfive commented:
r? @KodrAus

(rust-highfive has picked a reviewer for you, use r? to override)

usbalbin commented:
What do we want to call the feature gate? Would it make sense to reuse `const_intrinsic_copy` since copy_[to/from] are not much more than wrappers for the functions behind that?

RalfJung commented:
Cc @rust-lang/wg-const-eval 

> What do we want to call the feature gate for copy_to and copy_from? Would it make sense to reuse `const_intrinsic_copy` since copy_[to/from] are not much more than wrappers for the functions behind that?

Yeah that makes sense.

usbalbin commented on library/core/src/ptr/mod.rs:
```diff
@@ -452,15 +452,10 @@ pub const unsafe fn swap<T>(x: *mut T, y: *mut T) {
+#[rustc_const_unstable(feature = "const_swap", issue = "none")]
+pub const unsafe fn swap_nonoverlapping<T>(x: *mut T, y: *mut T, count: usize) {
+    // FIXME: Is it okay to remove these checks? The same checks were removed for
+    // ptr::copy_nonoverlapping
```
Do we want to keep these checks? The same checks were removed for `ptr::copy_nonoverlapping` when it was constified

RalfJung replied:
For `copy_nonoverlapping` we ended up removing them entirely also for performance reasons. I am not sure if those apply here.

usbalbin replied:
"Not sure" as in "Lets undo those changes for now so we can continue with the less controversial changes and maybe consider that for some other time"? Or just "Not sure" :)

RalfJung replied:
"Not sure" as in "let's see what others thing"... @oli-obk ?

oli-obk replied:
I'm for removing the checks. I don't see why `swap_nonoverlapping` should have them while `copy_nonoverlapping` doesn't have them.

usbalbin commented:
I have removed the `// FIXME` [mentioned above](https://github.com/rust-lang/rust/pull/83091#discussion_r593955521) and squashed some commits. Do you want me to squash all commits into one or keep it as it is, one per feature gate?

oli-obk commented on library/core/src/ptr/mod.rs:
```diff
@@ -394,7 +394,8 @@ pub const fn slice_from_raw_parts_mut<T>(data: *mut T, len: usize) -> *mut [T] {
 #[inline]
 #[stable(feature = "rust1", since = "1.0.0")]
-pub unsafe fn swap<T>(x: *mut T, y: *mut T) {
+#[rustc_const_unstable(feature = "const_swap", issue = "none")]
```
could you create and reference an issue for these?

oli-obk commented:
> Do you want me to squash all commits into one or keep it as it is, one per feature gate?

one commit per feature gate is great, thanks!

Once all unstable feature attributes have an issue instead of `none`, r=me

usbalbin commented:
@oli-obk 
> Once all unstable feature attributes have an issue instead of `none`, r=me

Ok, so...

@bors r=oli-obk

...I hope that is right :)

joshtriplett commented:
@bors r=oli-obk

bors commented:
:pushpin: Commit db9a53b5d7dcf2be60496a6ee19f03476d354344 has been approved by `oli-obk`

<!-- @bors r=oli-obk db9a53b5d7dcf2be60496a6ee19f03476d354344 -->
<!-- homu: {"type":"Approved","sha":"db9a53b5d7dcf2be60496a6ee19f03476d354344","approver":"oli-obk"} -->
